### PR TITLE
#306 ACL RoleMapper enhancements for stage dependent generic role handling

### PR DESCRIPTION
**Pull request by Kusig**
See https://jira.post.ch/browse/SDCISA-122

New Featureset:
- RoleMapper could be now cascaded by need
- RoleMapper could be used with environement properties
- The request headers are now filled in with the mapped roles if there are any.

### CODE REVIEW

File: gateleen-core/src/main/java/org/swisspush/gateleen/core/util/RoleExtractor.java [hunk 1]
```diff
@@ -16,23 +16,41 @@
  */
 public class RoleExtractor {
 
-    private static String groupHeader = "x-rp-grp";
-    private static String roleHeader = "x-roles";
+    public static String groupHeader = "x-rp-grp";
+    public static String roleHeader = "x-roles";
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-core/src/main/java/org/swisspush/gateleen/core/util/RoleExtractor.java]: Will do so.

File: gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleAuthorizer.java [hunk 2]
```diff
@@ -29,6 +29,8 @@
     private String deviceHeader = "x-rp-deviceid";
     private String userHeader = "x-rp-usr";
 
+    private String rolePrefix = "";
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleAuthorizer.java]: Would be possible but I do more like the explizit initialization which makes it clear what it is initialized whithout having to dig in the logic of the constructor. As well when changing something in the future or even adding another constructor, it is still properly initialized.

File: gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleAuthorizer.java [hunk 3]
```diff
@@ -141,6 +161,17 @@ private boolean isAuthorized(Set<String> roles, HttpServerRequest request) {
         return false;
     }
 
+    private void fillInNewRoleHeader(HttpServerRequest request,Set<String>roles)
+    {
+        StringBuffer roleHeader = new StringBuffer();
+        for (String role : roles)
+        {
+            if (roleHeader.length()>0) roleHeader.append(",");
+            roleHeader.append(this.rolePrefix + role.toLowerCase());
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleAuthorizer.java]: Yep, makes sense.

File: gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleMapperFactory.java [hunk 4]
```diff
@@ -21,8 +22,11 @@
 
     private static Logger LOGGER = LoggerFactory.getLogger(RoleMapperFactory.class);
 
-    RoleMapperFactory() {
+    private final Map<String, Object> properties;
+
+    RoleMapperFactory(Map<String, Object> properties) {
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleMapperFactory.java]: 1. JavaDoc added.
2. I asked myself the same question but it is treated like this within whole gateleen and I don't want to change the pattern just here.

**Comment by Kusig** [comment]
> Please also create a github issue describing the intentions for these changes as the information under https://jira.post.ch/browse/SDCISA-122 is not publicly accessible.

Exists already #306  and ist linked. Is there something wrong with that?

**Comment by mcweba** [comment]
> > Please also create a github issue describing the intentions for these changes as the information under https://jira.post.ch/browse/SDCISA-122 is not publicly accessible.
> 
> Exists already #306 and ist linked. Is there something wrong with that?

Sorry must have missed that. Nothing's wrong with that

File: gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/Authorizer.java [hunk 5]
```diff
@@ -52,16 +52,16 @@
 
     public static final Logger log = LoggerFactory.getLogger(Authorizer.class);
 
-    public Authorizer(Vertx vertx, final ResourceStorage storage, String securityRoot, String rolePattern) {
+    public Authorizer(Vertx vertx, final ResourceStorage storage, String securityRoot, String rolePattern, String rolePrefix, Map<String, Object> properties) {
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/Authorizer.java:55]: Ok, could do that.
Wasn't done by purpose in fact in order to force migrations and keep things clean :-)

File: gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleMapperHolder.java [hunk 6]
```diff
@@ -25,18 +25,24 @@
...
+     */
+    private final boolean continueMapping;
 
     /**
      * Holds the attributes needed to map roles. All params must be non null.
      *
      * @param pattern      The regular expression pattern to be applied to the incoming rules, must not be null
      * @param role         The resulting mapped role, must not be null
      * @param keepOriginal If true, the original role is kept and the mapped one added additionally
+     *
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleMapperHolder.java:43]: Added.

File: gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleAuthorizer.java [hunk 7]
```diff
@@ -141,6 +161,17 @@ private boolean isAuthorized(Set<String> roles, HttpServerRequest request) {
         return false;
     }
 
+    private void fillInNewRoleHeader(HttpServerRequest request,Set<String>roles)
+    {
+        StringBuffer roleHeader = new StringBuffer();
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleAuthorizer.java:167]: ok

File: gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/Authorizer.java [hunk 8]
```diff
@@ -52,16 +52,16 @@
...
         this.vertx = vertx;
         this.storage = storage;
         String aclRoot = UriBuilder.concatUriSegments(securityRoot, aclKey, "/");
         this.aclUriPattern = new PatternHolder(Pattern.compile("^" + aclRoot + "(?<role>.+)$"));
         this.userUriPattern = Pattern.compile(securityRoot + "user(\\?.*)?");
         this.roleMapperUriPattern = new PatternHolder(Pattern.compile("^" + UriBuilder.concatUriSegments(securityRoot, RoleMapper.ROLEMAPPER)));
         this.roleExtractor = new RoleExtractor(rolePattern);
-        this.roleMapper = new RoleMapper(storage, securityRoot);
-        this.roleAuthorizer = new RoleAuthorizer(storage, securityRoot, rolePattern, roleMapper);
+        this.roleMapper = new RoleMapper(storage, securityRoot,properties);
```
[the remark this answers is not in the document]
Comment (Kusig) [reply gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/Authorizer.java:74]: Applied CodeFormatter default but infact it is configured to be applied automatically ..

**Pull request merged**